src/main/scala/analysis/UtilMethods.scala

```diff
@@ -54,3 +55,76 @@ def evaluateExpression(exp: Expr, constantPropResult: Map[Variable, FlatElement[
       None
   }
 }
+
+def evaluateExpressionWithSSA(exp: Expr, constantPropResult: Map[RegisterVariableWrapper, Set[BitVecLiteral]]): Set[BitVecLiteral] = {
+  Logger.info(s"evaluateExpression: $exp")
+
+  def apply(op: (BitVecLiteral, BitVecLiteral) => BitVecLiteral, a: Set[BitVecLiteral], b: Set[BitVecLiteral]): Set[BitVecLiteral] =
+    val res = for {
+      x <- a
+      y <- b
+    } yield op(x, y)
+    res
+
+  def applySingle(op: BitVecLiteral => BitVecLiteral, a: Set[BitVecLiteral]): Set[BitVecLiteral] =
+    val res = for {
+      x <- a
+    } yield op(x)
+    res
+
+
+  exp match {
+    case binOp: BinaryExpr =>
+      val lhs = evaluateExpressionWithSSA(binOp.arg1, constantPropResult)
+      val rhs = evaluateExpressionWithSSA(binOp.arg2, constantPropResult)
+
+      (lhs, rhs) match {
+        case (l: Set[BitVecLiteral], r: Set[BitVecLiteral]) =>
+          binOp.op match {
+            case BVADD => apply(BitVectorEval.smt_bvadd, lhs, rhs)
+            case BVSUB => apply(BitVectorEval.smt_bvsub, lhs, rhs)
+            case BVASHR => apply(BitVectorEval.smt_bvashr, lhs, rhs)
+            case BVCOMP => apply(BitVectorEval.smt_bvcomp, lhs, rhs)
+            case BVCONCAT => apply(BitVectorEval.smt_concat, lhs, rhs)
+            case BVLSHR => apply(BitVectorEval.smt_bvlshr, lhs, rhs)
+            case BVSHL => apply(BitVectorEval.smt_bvshl, lhs, rhs)
+            case BVOR => apply(BitVectorEval.smt_bvor, lhs, rhs)
+            case _ => throw new RuntimeException("Binary operation support not implemented: " + binOp.op)
+          }
+      }
+    case extend: ZeroExtend =>
+      val result = evaluateExpressionWithSSA(extend.body, constantPropResult)
+      applySingle(BitVectorEval.smt_zero_extend(extend.extension, _: BitVecLiteral), result)
+    case e: Extract =>
+      val result = evaluateExpressionWithSSA(e.body, constantPropResult)
+      applySingle(BitVectorEval.boogie_extract(e.end, e.start, _: BitVecLiteral), result)
+    case variable: Variable =>
+      val registerVariableWrapper = RegisterVariableWrapper(variable)
+      constantPropResult.collect({
+        case (k, v) if k == registerVariableWrapper => v
+      }).flatten.toSet
+      constantPropResult(registerVariableWrapper)
+    case b: BitVecLiteral => Set(b)
+    case _ => //throw new RuntimeException("ERROR: CASE NOT HANDLED: " + exp + "\n")
+      Set()
+  }
+}
+
+def unwrapExpr(expr: Expr): ListBuffer[Expr] = {
+  val buffers = ListBuffer[Expr]()
+  expr match {
+    case e: Extract => unwrapExpr(e.body)
+    case e: SignExtend => unwrapExpr(e.body)
+    case e: ZeroExtend => unwrapExpr(e.body)
+    case repeat: Repeat => unwrapExpr(repeat.body)
+    case unaryExpr: UnaryExpr => unwrapExpr(unaryExpr.arg)
+    case binaryExpr: BinaryExpr =>
+      unwrapExpr(binaryExpr.arg1)
+      unwrapExpr(binaryExpr.arg2)
+    case memoryLoad: MemoryLoad =>
+      buffers.addOne(memoryLoad)
+      unwrapExpr(memoryLoad.index)
+    case _ =>
+  }
+  buffers
+}
```

Comment:
What is this supposed to do? My impression is that it will either return a ListBuffer containing a MemoryLoad if the input Expr is a MemoryLoad, or an empty ListBuffer if Expr is not a MemoryLoad, and the recursive calls will do nothing?

Reply:
@yousifpatti

Reply:
This is supposed to be recursively called and returns a List containing MemoryLoads identified in the expression as it may contain multiple MemoryLoads
I will fix this

Reply:
It is safe to assume that an Expr does not contain multiple MemoryLoads recursively

Reply:
but you also need to actually apply the semantics of anything the MemoryLoad is contained within

Reply:
MRA is only concerned about collecting the regions being loaded and not the actual semantics of the expression as it is not evaluated as a whole. Collecting those regions will provide a more fine-grained MemoryModelMap.
The IR language allows for a binary expression to contain multiple MemoryLoads for examples so you cannot assume that multiple loads will not exist in a single expression